File: Iter2/src/StudentCourse.java
/**
 * The StudentCourse class that holds information and methods regarding the relationship between students and courses.
 */
public class StudentCourse {
    /**
     * Student's course grade [0, 100]
     */
    private int grade;
    /**
     * Student's course point [0, 4]
     */
    private double point;
    /**
     * Letter grade calculated from the course grade
     */
    private String letterGrade;
    /**
     * Main department course (lectures)
     */
    private DepartmentCourse mainCourse;
    /**
     * Student is passed or not
     */
    private boolean passed;
    /**
     * In which semester the student took the course
     */
    private int registeredSemester;
    /**
     * Registration warning messages
     */
    private final String[] StudentRegistrationMessages = {
            "Student is not eligible for course %s because prerequisite %s did not meet.",
            "Advisor did not approve take of %s because he/she has another course at course time.",
            "Advisor did not approve take of %s because he/she reached credit take limit (semesters: %s).",
            "Student is could not take %s because no capacity left.",
            "Student is could not take this lab section %s because no capacity left.",
            "Advisor did not approve taking %s because he/she has another course at that time.",
            "Student cannot take %s in this semester because the course belongs to another semester.",
            "Student cannot take %s because he/she did not meet minimum completed credits.",
    };

    private final String succesful = "Student successfully registered to %s";

    public StudentCourse(DepartmentCourse course, int grade, int semester) {
        this.mainCourse = course;
        setGrade(grade);
        this.point = getPoint();
        this.letterGrade = getLetterGrade();
        this.registeredSemester = semester;
    }

    public void checkRules(Student student, DepartmentCourse course) {

        Department department = student.getDepartment();
        String message = "";
        String succesfulMessage = "";
        String courseIdName = course.getCourseId() + " " + course.getCourseName();
        int semester = course.getSemester();


        if (semester == 0 && course instanceof ElectiveCourse) {
            semester = student.getElectiveSemester((ElectiveCourse) course);
        }

        if (!student.getFailedPrerequisites(course).isEmpty()) { // Check prerequisites
            message = String.format(StudentRegistrationMessages[0], courseIdName, student.getFailedPrerequisites(course).get(0).getCourseName());
            student.addRegistrationMessage(message);
            department.addCourseRegistrationMessage(0, course);
        }
        if (student.getTimetable().courseHasCollision(course)) { // Check collisions
            message = String.format(StudentRegistrationMessages[1], courseIdName);
            student.addRegistrationMessage(message);
            department.addCourseRegistrationMessage(1, course);
        }
        if (student.getTakenCredits(semester) + course.getCredits() > department.getMaximumCredits()) { // Check credits
            message = String.format(StudentRegistrationMessages[2], courseIdName, course.getSemester());
            student.addRegistrationMessage(message);
            department.addCourseRegistrationMessage(2, course);
        }

        if (course instanceof ElectiveCourse && !((ElectiveCourse) course).hasCapacity()) { // Check capacity
            message = String.format(StudentRegistrationMessages[3], courseIdName);
            student.addRegistrationMessage(message);
            department.addCourseRegistrationMessage(3, course);
        }

        if (!course.getLabSections().isEmpty()) { //Check lab sections
            int random = (int) (Math.random() * course.getLabSections().size());
            LabCourse randomLabSection = course.getLabSections().get(random);


            if (!randomLabSection.hasCapacity()) { // Check lab section capacity
                message = String.format(StudentRegistrationMessages[4], randomLabSection.getCourseId());
                student.addRegistrationMessage(message);
                department.addCourseRegistrationMessage(4, randomLabSection);
            }

            if (student.getTimetable().courseHasCollision(randomLabSection)) { // Check lab section collisions
                message = String.format(StudentRegistrationMessages[5], randomLabSection.getCourseId());
                student.addRegistrationMessage(message);
                department.addCourseRegistrationMessage(5, randomLabSection);
            }
            if (randomLabSection.hasCapacity() && !student.getTimetable().courseHasCollision(randomLabSection) && message.isEmpty()) {
                student.getTimetable().assignCourse(randomLabSection);
                randomLabSection.addStudent(student);
            }
        }

        if (course.getSemester() > student.getSemester()) { // Check semester
            message = String.format(StudentRegistrationMessages[6], courseIdName);
            student.addRegistrationMessage(message);
            department.addCourseRegistrationMessage(6, course);
        }

        if (course.getMinCompletedCredits() > 0 && student.getCompletedCredits(semester - 1) < course.getMinCompletedCredits()) { // Check completed credits
            message = String.format(StudentRegistrationMessages[7], courseIdName, course.getSemester());
            student.addRegistrationMessage(message);
            department.addCourseRegistrationMessage(7, course);
        }

        if (message.isEmpty()) { //check if registration is successful
            student.registerCourse(course, RandomStudentGenerator.randomGrade(), semester);
            succesfulMessage = String.format(succesful, courseIdName);
            student.addRegistrationMessage(succesfulMessage);
        }

    }

    public void setGrade(int grade) {
        this.grade = grade;
        setLetterGrade(grade);
        if (point < 1.5) {
            passed = false;
        }
    }

    public String getLetterGrade() {
        return letterGrade;
    }

    public void setLetterGrade(int grade) {
        if (grade >= 90) {
            letterGrade = "AA";
            point = 4;
        } else if (grade >= 85) {
            letterGrade = "BA";
            point = 3.5;
        } else if (grade >= 80) {
            letterGrade = "BB";
            point = 3;
        } else if (grade >= 75) {
            letterGrade = "CB";
            point = 2.5;
        } else if (grade >= 65) {
            letterGrade = "CC";
            point = 2;
        } else if (grade >= 55) {
            letterGrade = "DC";
            point = 1.5;
        } else if (grade >= 50) {
            letterGrade = "DD";
            point = 1;
        } else if (grade >= 45) {
            letterGrade = "FD";
            point = 0.5;
        } else {
            letterGrade = "FF";
            point = 0;
        }
    }


    public DepartmentCourse getMainCourse() {
        return mainCourse;
    }


    public int getRegisteredSemester() {
        return registeredSemester;
    }

    public double getPoint() {
        return point;
    }


    public int getGrade() {
        return grade;
    }

    public boolean isPassed() {
        return passed;
    }
}

File: Iter3/src/student_course.py
import random
from typing import List

import course

import student as student
from department_course import DepartmentCourse


class StudentCourse:
    __grade: int
    __point = 4.00 
    __letterGrade: str
    __mainCourse: type(DepartmentCourse)
    __passed = True
    __registeredSemester: int
    __registrationMessages: List[str] = [
        "Student is not eligible for course {} because prerequisite {} did not met.",
        "Advisor did not approve take of {} because he/she has another course at course time.",
        "Advisor did not approve take of {} because he/she reached credit take limit (semesters: {}).",
        "Student is could not take {} because no capacity left.",
        "Student is could not take this lab section {} because no capacity left.",
        "Advisor did not approve taking {} because he/she has another course at that time.",
        "Student cannot take {} in this semester because the course belongs to another semester.",
        "Student cannot take {} because he/she did not met minimum completed credits."]

    def __init__(self, arg_course, grade, semester):
        self.__mainCourse = arg_course
        self.set_grade(grade)
        self.__registeredSemester = semester

    def check_rules(self, student: student, arg_course: course.DepartmentCourse):
        new_department = student.get_department()
        message = ""
        course_id_name = arg_course.get_course_id() + " " + arg_course.get_course_name()
        semester = arg_course.get_semester()

        if semester == 0 and isinstance(arg_course, course.ElectiveCourse):
            semester = student.get_elective_semester(arg_course)

        if arg_course.get_semester() > student.get_semester():
            message = self.__registrationMessages[6].format(course_id_name)
            student.add_registration_message(message)
            new_department.add_course_registration_message(6, arg_course)

        if len(student.get_failed_prerequisites(arg_course)) > 0:
            message = self.__registrationMessages[0].format(course_id_name,
                                                            student.get_failed_prerequisites(arg_course)(
                                                                0).get_course_name())
            student.add_registration_message(message)
            new_department.add_course_registration_message(0, arg_course)

        if student.get_time_table().course_has_collision(arg_course):
            message = self.__registrationMessages[1].format(course_id_name)
            student.add_registration_message(message)
            new_department.add_course_registration_message(1, arg_course)

        if student.get_taken_credits(semester) + arg_course.get_credits() > new_department.get_maximum_credits():
            message = self.__registrationMessages[2].format(course_id_name, arg_course.get_semester())
            student.add_registration_message(message)
            new_department.add_course_registration_message(2, arg_course)

        if arg_course.get_min_completed_credits() > 0 and student.get_completed_credits(
                semester - 1) < arg_course.get_min_completed_credits():
            message = self.__registrationMessages[7].format(course_id_name, arg_course.get_semester())
            student.add_registration_message(message)
            new_department.add_course_registration_message(7, arg_course)

        if isinstance(arg_course, course.ElectiveCourse) and not arg_course.has_capacity():
            message = self.__registrationMessages[3].format(course_id_name)
            student.add_registration_message(message)
            new_department.add_course_registration_message(3, arg_course)

        if len(arg_course.get_lab_sections()) > 0:
            random_lab_section: course.LabCourse
            random_lab_section = random.choice(arg_course.get_lab_sections())

            if not random_lab_section.has_capacity():
                message = self.__registrationMessages[4].format(random_lab_section.get_course_id())
                student.add_registration_message(message)
                new_department.add_course_registration_message(4, random_lab_section)

            if student.get_time_table().course_has_collision(random_lab_section):
                message = self.__registrationMessages[5].format(random_lab_section.get_course_id())
                student.add_registration_message(message)
                new_department.add_course_registration_message(5, random_lab_section)

            if random_lab_section.has_capacity() and not student.get_time_table().course_has_collision(
                    random_lab_section) and len(message) == 0:
                student.get_time_table().assign_course(random_lab_section)
                random_lab_section.add_student(student)

        else:
            student.register_course(arg_course, RandomStudentGenerator.random_grade(), semester)

    def set_grade(self, grade: int):
        self.__grade = grade
        self.set_letter_grade(grade)
        if self.__point < 1.5:
            self.__passed = False

    def get_letter_grade(self):
        return self.__letterGrade

    def set_letter_grade(self, grade: int):
        if grade >= 90:
            self.__letterGrade = "AA"
            self.__point = 4.00
        elif grade >= 85:
            self.__letterGrade = "BA"
            self.__point = 3.5
        elif grade >= 80:
            self.__letterGrade = "BB"
            self.__point = 3
        elif grade >= 75:
            self.__letterGrade = "CB"
            self.__point = 2.5
        elif grade >= 65:
            self.__letterGrade = "CC"
            self.__point = 2
        elif grade >= 55:
            self.__letterGrade = "DC"
            self.__point = 1.5
        elif grade >= 50:
            self.__letterGrade = "DD"
            self.__point = 1
        elif grade >= 45:
            self.__letterGrade = "FD"
            self.__point = 0.5
        else:
            self.__letterGrade = "FF"
            self.__point = 0

    def get_main_course(self):
        return self.__mainCourse

    def get_registered_semester(self):
        return self.__registeredSemester

    def get_point(self):
        return self.__point

    def get_grade(self):
        return self.__grade

    def is_passed(self):
        return self.__passed
